scala/tempo/src/main/scala/com/databrickslabs/tempo/resample.scala

```diff
@@ -0,0 +1,112 @@
+package com.databrickslabs.tempo
+
+import org.apache.spark.sql.functions._
+import TSDF._
+
+/**
+  * The following object contains methods for resampling (up to millions of time series in parallel).
+  */
+object resample {
+
+// define global frequency options
+val SEC = "sec"
+val MIN = "min"
+val HR = "hr"
+
+// define global aggregate function options for downsampling
+val CLOSEST_LEAD = "closest_lead"
+val MIN_LEAD = "min_lead"
+val MAX_LEAD = "max_lead"
+val MEAN_LEAD = "mean_lead"
+
+val allowableFreqs = List(SEC, MIN, HR)
+
+  /**
+    *
+    * @param tsdf - TSDF object as input
+    * @param freq - frequency at which to upsample
+    * @return - return a TSDF with a new aggregate key (called agg_key)
+    */
+  def __appendAggKey(tsdf : TSDF, freq : String) : TSDF = {
+    var df = tsdf.df
+    checkAllowableFreq(freq)
+
+    // compute timestamp columns
+    val sec_col = second(col(tsdf.tsColumn.name))
+    val min_col = minute(col(tsdf.tsColumn.name))
+    val hour_col = hour(col(tsdf.tsColumn.name))
+    var agg_key = col(tsdf.tsColumn.name)
+
+    if (freq == SEC) {
+      agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lpad(min_col, 2, "0"), lit(":"), lpad(sec_col, 2, "0")).cast("timestamp")
+    } else if (freq == MIN) {
+        agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lpad(min_col, 2, "0"), lit(":"), lit("00")).cast("timestamp")
+    } else if (freq == HR) {
+        agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lit("00"), lit(":"), lit("00")).cast("timestamp")
+    }
+
+    df = df.withColumn("agg_key", agg_key)
+    return TSDF(df, tsColumnName= tsdf.tsColumn.name, partitionColumnNames = tsdf.partitionCols.map(x => x.name): _*)
+  }
+
+  /**
+    *
+    * @param tsdf - input TSDF object
+    * @param freq - aggregate function
+    * @param func - aggregate function
+    * @param metricCols - columns used for aggregates
+    * return - TSDF object with newly aggregated timestamp as ts_col with aggregated values
+    */
+  def rs_agg(tsdf : TSDF, freq : String, func : String, metricCols : Option[List[String]] = None) : TSDF =  {
```

Comment:
I feel like in practice this can lead to major skew challenges from a perf standpoint. I would prob lean on the stats package to do some sampling and stats calcs to see if salting would be necessary and add a generalized salting package to tempo to enable in key locations based on stats.